### DISCUSSION TITLE
ImportError: libmysqlclient.so.20

### DISCUSSION THREAD

**Discussion by Juvelir97** [apache/airflow#29419] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Hello,

Trying to connect Airflow-2.4.3 to Mysql 8.0.29-1 Database 
Using the string inside airflow.cfg 
sql_alchemy_conn = mysql+mysqldb://USER:PASS@HOST:3306/airflow

running airflow commands (e,g, airflow db init or  airflow config get-value database sql_alchemy_conn)we are facing the issue:
Traceback (most recent call last):
  File "/opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/MySQLdb/__init__.py", line 18, in <module>
    from . import _mysql
ImportError: libmysqlclient.so.20: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/mapr/airflow/airflow-2.4.3/bin/airflow", line 4, in <module>
    from airflow.__main__ import main
  File "/opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/airflow/__init__.py", line 46, in <module>
    settings.initialize()
  File "/opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/airflow/settings.py", line 571, in initialize
    configure_adapters()
  File "/opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/airflow/settings.py", line 436, in configure_adapters
    import MySQLdb.converters
  File "/opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/MySQLdb/__init__.py", line 24, in <module>
    version_info, _mysql.version_info, _mysql.__file__
NameError: name '_mysql' is not defined

Can you please assist?
Thanks in advance

### Operating System

RHEL 8.4

### Deployment

Other

### Deployment details

Was deployed as part of Mapr cluster

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
This is not an airflow issue. Converted to discussion. This is a problem with your environment not having mysql client system libraries installed. Please install them as appropriate. You will likely find right way at mysql installation pages and you can google similar issues people have..

**Reply by Juvelir97** [reply]
Thank you for a quick reply.
We have two another test Ubuntu 20.04 clusters with identical configuration and the only difference is the airflow version(2.2.1 and 2.4.3).
2.2.1 version works fine
2.4.3 falling with the same error.
Both don't have a lib libmysqlclient.so.20 but libmysqlclient.so.21 instead

**Reply by Juvelir97** [reply]
Also 

Running this on the problem cluster :
$ grep -R 'libmysqlclient.so.20' /opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/* 2>/dev/null
Binary file /opt/mapr/airflow/airflow-2.4.3/build/env/lib/python3.9/site-packages/MySQLdb/_mysql.cpython-39-x86_64-linux-gnu.so matches

Running this on the working cluster
grep -R 'libmysqlclient.so.20' /opt/mapr/airflow/airflow-2.2.1/build/env/lib/python3.9/site-packages/* 2>/dev/null
Return nothing.

on cluster with 2.2.1 airflow even there is no dir  MySQLdb under site-packages dir

**Reply by potiuk** [reply]
I am not in a position to look at your system and how it was installed/prepared - the dependencies uses (mysql-related) might have different requirements and you should make sure you will get the right installation.  There is no guarantee that the same dependencies are used in old an new version. Which driver is used by sqlalchemy, depends on your SQL Alchemy URL that you chose. And then - each driver might have different requirements. It looks like you should read the related docs and possibly ask question in the right driver/version you use. 

Googling similar problems is usually a good idea as well. This is what I typically do when I have similar problems and I strongly advice you to look for similar problems and possible solutions.

This is also the consequence of using "local installation". If you go that path, then indeed you are responsible to make sure that your environment is well prepared and that drivers you use have the right libraries etc. There is no way we can help here in Airlfow as a general "problem solving". But we are also publishing Docker Containers that are based on specific versions of the base images (debian in this case) and have specific requirements installed to get Airflow works (and are tested in our CI). https://airflow.apache.org/docs/docker-stack/index.html. There - problems like this are far less likely.

Docker container approach is specifically targeted to prevent similar problems you have - by choosing local installation you chose a more difficult path, where you have to - exactly - fight those kind of problems on your own I am afraid (and googling might help to see if others have some ready recipes for it).
